File: backend/routers/reports.py
from fastapi import APIRouter, Query
from collections import defaultdict
from google_sheets_service import get_all_records
from models.report import Student

router = APIRouter()

@router.post("/student")
def student_report(student: Student):
    orders = get_all_records("Orders")
    prices = get_all_records("Prices")
    print("Orders:", orders)
    # print("Prices:", prices)
    print(student)
    # Filter orders for the student, school, and class
    filtered = [o for o in orders if o.get("student_name") == student.student_name and o.get("school_name") == student.school_name and o.get("class") == student.class_name]
    if not filtered:
        return {"detail": "No report found"}
    items = []
    total_amount = 0
    for o in filtered:
        item_name = o.get("item_name")
        size = o.get("size") if "size" in o else o.get("item_size")
        qty = int(o.get("qty", 1))
        # Find price for this item/size/gender/school
        price_row = next((p for p in prices if p.get("item_name") == item_name and p.get("item_size", p.get("size")) == size and p.get("school_name") == student.school_name), None)
        price = float(price_row["price"]) if price_row and "price" in price_row else 0
        item_total = qty * price
        total_amount += item_total
        items.append({
            "item_name": item_name,
            "size": size,
            "qty": qty,
            "price": price
        })
    return {
        "student_name": str(student.student_name),
        "school_name": str(student.school_name),
        "class_name": str(student.class_name),
        "items": items,
        "total_amount": total_amount
    }

File: frontend/js/reports.js

const API_BASE = "http://localhost:8000";

async function loadSchoolsForReport() {
    const res = await fetch(`${API_BASE}/schools`);
    if (!res.ok) return;
    const schools = await res.json();
    const schoolSelect = document.getElementById("schoolName");
    schoolSelect.innerHTML = '<option value="">Select School</option>';
    schools.forEach(s => {
        const opt = document.createElement("option");
        opt.value = s.school_name || s.name || s.school; // support different key names
        opt.textContent = s.school_name || s.name || s.school;
        schoolSelect.appendChild(opt);
    });
}

document.addEventListener("DOMContentLoaded", loadSchoolsForReport);

document.getElementById("studentReportForm").addEventListener("submit", async (e) => {
    e.preventDefault();
    const studentName = document.getElementById("studentName").value.trim();
    const schoolName = document.getElementById("schoolName").value.trim();
    const className = document.getElementById("className").value.trim();
    // Send parameters as JSON in POST request
    const res = await fetch(`${API_BASE}/reports/student`, {
        method: "POST",
        headers: {"Content-Type": "application/json"},
        body: JSON.stringify({
            student_name: studentName,
            school_name: schoolName,
            class_name: className
        })
    });
    if (!res.ok) {
        document.getElementById("studentReportResult").innerHTML = `<div class='alert alert-danger'>No report found for the given details.</div>`;
        return;
    }
    const report = await res.json();
    let html = `<h4>Student: ${report.student_name}</h4>`;
    html += `<table class='table table-bordered'><thead><tr><th>Item Name</th><th>Size</th><th>Qty</th><th>Price</th><th>Total Price</th></tr></thead><tbody>`;
    let total = 0;
    report.items.forEach(item => {
        const itemTotal = item.qty * item.price;
        total += itemTotal;
        html += `<tr><td>${item.item_name}</td><td>${item.size}</td><td>${item.qty}</td><td>₹${item.price}</td><td>₹${itemTotal}</td></tr>`;
    });
    html += `</tbody></table>`;
    html += `<h5>Total Bill Amount: ₹${total}</h5>`;
    document.getElementById("studentReportResult").innerHTML = html;
});




